dockstore-webservice/src/main/java/io/dockstore/webservice/core/Version.java

```diff
@@ -237,6 +237,11 @@
     @ApiModelProperty(value = "The user-specified files for the version.")
     private List<String> userFiles = new ArrayList<>();
 
+    @Column(columnDefinition = "varchar")
+    @Convert(converter = UserImageReferencesConverter.class)
+    @ApiModelProperty(value = "The user-specified image references for the version.")
+    private List<String> userImageReferences = new ArrayList<>();
```

Comment:
Here's the my understanding/intentions:
The `userImageReferences` field contains a list of references to images (on quay or wherever).
The `images` field contains the images have been confirmed/checksummed during the snapshot process (or loaded during tool registration/refresh).  Each constituent `Image` doesn't exist until one of those happens (please correct me if I'm wrong).

So:
1. No, IMHO, an image reference and the `Image` object are different things.  A reference contains the identifying information used to retrieve an `Image`.  One is an identifier (much like those in the toolTable result from `getContent`) and the other is the thing identified.  Also, note that we may not have access to the referenced "kernel image", it may not be in a public registry.  The reference will be passed to the JupyterHub server, and it will retrieve the image.  We may not be able to retrieve it ourselves (snapshotting will fail in that case).
1a. During the notebooks PR that included "user paths", there was strong opposition to not storing the user input (the paths) in the Version.  This solution stores the user input in the Version.
2. It's intentionally not just for notebooks. Rather than hardcode a "kernel image", it's designed so it might be useful for some other type of entry someday.
3. Given the current architecture and state of the code, I strongly recommend against creating `NotebookVersion` unless/until we absolutely have to, and jamming everything into `WorkflowVersion`.  Creating a `NotebookVersion` would be lots of the extra code and disruption, see the diffs for a comment that alludes to a reason about why we punted on a separate class for services. The fact that we don't have a different `Version` entity for each type of entry strongly hints at this.

I agree that my proposed solution is not pure/perfect, but given the conditions on the ground, I believe it to be a pragmatic choice that allows us to reach our goals in a reasonable amount of time.

Reply:
Also, note that the addition of snapshotting is easy, we simply process the references from `userImageReferences` like we do  (and in addition to) the references from the `getContent` tool table result.

Reply:
> If I understand your explanation correctly, userImageReferences is like Image+imageregistry + Image+tag + Image+repository but not like Image.size, Image.checksums yet. You're proposing that the user provide the image reference in .dockstore.yml (good IMHO), then store the reference in userImageReferences at github app release time, until a snapshot at which time Image will be populated, (a bit redundant?).

and

> Also, note that the addition of snapshotting is easy, we simply process the references from userImageReferences like we do (and in addition to) the references from the getContent tool table result.

@svonworl , in the current, pre-feedback implementation, are you proposing populating the images table when snapshotting, or would snapshotting of userImageReferences be done another way? I'm not getting the reference/similarity to `getContent` -- which at a skim, I'm not sure we set when we snapshot (we should).